File: project/src/main/java/com/expedia/ui/javademo/FeedItem.java
package com.expedia.ui.javademo;

/**
 * Created by nbirla on 15/11/17.
 */

public class FeedItem<D> {

    enum ExpandState {
        NONE, EXPANDED, EXPANDING, COLLAPSING, COLLAPSED
    }

    private String id;
    private String priorityType;
    private String viewType;
    private D d;

    private ExpandState state = ExpandState.COLLAPSED;

    public FeedItem(String viewType, D d){
        this.viewType = viewType;
        this.d = d;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getPriorityType() {
        return priorityType;
    }

    public void setPriorityType(String priorityType) {
        this.priorityType = priorityType;
    }

    public String getViewType() {
        return viewType;
    }

    public D getBindingData() {
        return d;
    }

    public ExpandState getExpandState() {
        return state;
    }

    public void setExpandState(ExpandState expandState) {
        this.state = expandState;
    }
}

File: project/src/main/java/com/expedia/ui/recyclerview/FeedItem.kt
package com.expedia.ui.recyclerview

/**
 * Created by nbirla on 15/11/17.
 */
class FeedItem<D>(val viewType: String, val d: D) {

    enum class ExpandState {
        NONE, EXPANDED, EXPANDING, COLLAPSING, COLLAPSED
    }

    constructor(feeditem :FeedItem<D>) : this(feeditem.viewType, feeditem.getBindingData()) {
        this.state = feeditem.state

        if(feeditem.id != null) {
            this.id = feeditem.id
        }

        if(feeditem.priorityType != null){
            this.priorityType = feeditem.priorityType
        }
    }

    private var id: String? = null
    private var priorityType: String? = null

    private var state : ExpandState = ExpandState.COLLAPSED

    fun getId(): String? {
        return id
    }

    fun setId(id: String) {
        this.id = id
    }

    fun getBindingData(): D {
        return d
    }

    fun getFeedItemViewType(): String {
        return viewType
    }

    fun getPriorityType(): String? {
        return priorityType
    }

    fun setPriorityType(priorityType: String) {
        this.priorityType = priorityType
    }

    fun getExpandState(): ExpandState {
        return state
    }

    fun setExpandState(state: ExpandState) {
        this.state = state
    }
}